### PR TITLE
Move the init of static variables to library load time

### LINKED ISSUE #1593: Possible race condition in FPFH feature, OMP implementation
I am performing point cloud registration with a feature based methods ISS3D, FPFH and computing normals. For all these method I used the OMP version of them. But I found that with OpenMP enabled (4 threads) I get inconsistent results between the runs on the registration. Further looking into OpenMP it does say some implementation can suffer from "race condition".

I've done multiple runs and tests with OMP on and off. With ooff always gives me same result but with OMP results are inconsistent. 

Has anyone else seen this happen?

### PR DESCRIPTION
**Pull request by kunaltyagi**
Possibly fixes #1593 (race condition in OMP implementation of SHOT)

### CODE REVIEW

**Comment by kunaltyagi** [comment]
Wow, seems like apple-clang doesn't keep the lambdas unique

**Review by vseib — COMMENTED** [review]
Dear all, 
this looks like a valid solution to #1593, thanks @kunaltyagi
I am not sure if I am allowed to approve a pull request, therefore I just create a review with a general comment.
Best,
Viktor

**Comment by mvieth** [comment]
@kunaltyagi What do you think about moving the `RGB2LAB` functionality to the common module? There is a function in [gicp6d](https://github.com/PointCloudLibrary/pcl/blob/master/registration/src/gicp6d.cpp#L45) that basically does the same (without LUT).

**Comment by kunaltyagi** [comment]
Sounds like a good idea. I'll take a deeper look either tomorrow or after the weekend.

**Comment by kunaltyagi** [comment]
@vseib @mvieth PTAL This should still remove the race condition 🤞🏾 (static vars are specially protected by the standard), while allowing code reuse.

Added a bunch of references as well which I hunted down while making sure I wasn't messing up anything major

**Comment by kunaltyagi** [comment]
@larshg Need help on windows 😄

**Comment by larshg** [comment]
It seems MSVC has a bug with constexpr and lambda - see [here](https://developercommunity.visualstudio.com/t/constexpr-compilation-bug-when-used-inside-lambda/250662).

Setting them to just static instead of constexpr, then it works - not sure if that is sufficient?

Alternatively, you could add an additional definition of `constexpr const std::size_t size = 1 << bits;` within the lambda.

**Comment by kunaltyagi** [comment]
Thanks @larshg. That worked wonders

**Comment by larshg** [comment]
> Thanks @larshg. That worked wonders

Great - and its the same for `discretizations` 😅

**Review by larshg — APPROVED** [review]

**Comment by kunaltyagi** [comment]
I'll have to check this error on Windows before merging:
```
test\registration\test_registration.cpp(619): error: Expected: (reg.getFitnessScore ()) < (0.003), actual: 0.00387298 vs 0.003
```

**Comment by kunaltyagi** [comment]
I'm unable to test why Windows is having such a large error in registration. On linux machines, the error is approx 4e-5, much lower than on Windows (reducing the LUT size used in GICP6D increases error, but going to 1024 increases it to only 4.5e-5)

Any ideas? Is this due to difference in c math lib?

**Comment by mvieth** [comment]
> I'm unable to test why Windows is having such a large error in registration. On linux machines, the error is approx 4e-5, much lower than on Windows (reducing the LUT size used in GICP6D increases error, but going to 1024 increases it to only 4.5e-5)
> 
> Any ideas? Is this due to difference in c math lib?

Perhaps because the second LUT is of float type, not of double? Not really a good explanation, but something to test out. Else you could try to replace the second LUT with a normal computation as it was before to check if that is the problem.
BTW I originally had something different in mind, maybe something like a simple class that creates the LUTs in the constructor and has a `RGB2LAB` function. SHOT and GICP6D could each have such an object as a class member, and use the conversion function when needed. I think with the current solution, in `RGB2Lab` in GICP6D, the LUTs are computed every time the function is called, aren't they? I think that would ultimately be worse than without LUTs.
But if you don't want to implement such a class and only focus on fixing the race condition in this PR, that would be fine, too.

**Comment by kunaltyagi** [comment]
> the LUTs are computed every time the function is called, aren't they?

This shouldn't be the case. The LUT are static variables inside the function. They are created on the first function call, and in a thread-safe manner (guarantee of language). Once init, they are returned safely.

Re: adding a class for easy conversion
IMO, the class for converting between colors should be left for another time, since colors aren't something that PCL specializes in either.

Re: float vs double
I'll push an update with double LUT since it's a plausible explanation. :) Let's see.

File: registration/src/gicp6d.cpp [hunk 1]
```diff
@@ -46,65 +47,39 @@ RGB2Lab(const Eigen::Vector3i& colorRGB)
...
-  Z = R * 0.0193 + G * 0.1192 + B * 0.9505;
-
-  // *= 100.0 including:
-  X /= 0.95047; // 95.047;
-  //    Y /= 1;//100.000;
-  Z /= 1.08883; // 108.883;
+  // normalize X, Y, Z with tristimulus values for Xn, Yn, Zn
+  float f[3] = {static_cast<float>(X), static_cast<float>(Y), static_cast<float>(Z)};
+  f[0] /= 0.95047;
+  f[1] /= 1;
```
Comment (mvieth) [inline registration/src/gicp6d.cpp:66]: Can we assume that every compiler is smart enough to not actually compute this, or should this be a comment to be safe?

**Review by mvieth — APPROVED** [review]

File: registration/src/gicp6d.cpp (hunk 1 above)
Comment (kunaltyagi) [reply registration/src/gicp6d.cpp:66]: It's a 100% NO-OP, at least for the compilers we support.

At least for the oldest compilers on godbolt, even those that we don't support: [Clang 3.0.0, GCC 4.1.2 and MSVC 19.14](https://godbolt.org/z/d7v44fWbb)
Releases:
* GCC 4.1.2: 2007, Feb 13
* Clang 3.0.0: 2011, Dec 1
* MSVC 19.14: 2018, May 7

Comment (mvieth) [reply registration/src/gicp6d.cpp:66]: Alright :+1:

**Comment by mvieth** [comment]
Ok, from my side, this is good to merge :+1:

**Pull request merged**